arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -0,0 +1,295 @@
+  return out;
+}
+}  // namespace graph
+}  // namespace arangodb
+
+/*
+IndexAccessor::IndexAccessor(transaction::Methods::IndexHandle idx,
+                             aql::AstNode* condition, std::optional<size_t> memberToUpdate)
+    : _idx(idx), _indexCondition(condition), _memberToUpdate(memberToUpdate) {}
+
+aql::AstNode* IndexAccessor::getCondition() const { return _indexCondition; }
+transaction::Methods::IndexHandle IndexAccessor::indexHandle() const {
+  return _idx;
+}
+std::optional<size_t> IndexAccessor::getMemberToUpdate() const {
+  return _memberToUpdate;
+}*/
+
+ClusterProvider::Step::Step(VertexType v)
+    : _vertex(v), _edge(), _fetched(false) {}
+
+ClusterProvider::Step::Step(VertexType v, EdgeDocumentToken edge, size_t prev)
+    : BaseStep(prev), _vertex(v), _edge(std::move(edge)), _fetched(false) {}
+
+ClusterProvider::Step::~Step() = default;
+
+VertexType const& ClusterProvider::Step::Vertex::getID() const {
+  return _vertex;
+}
+EdgeDocumentToken const& ClusterProvider::Step::Edge::getID() const {
+  return _token;
+}
+bool ClusterProvider::Step::Edge::isValid() const {
+  // TODO: Check isValid, when it is actually valid here in cluster case?
+  if (getID().vpack() != nullptr) {
+    return true;
+  }
+  return false;
+};
+
+void ClusterProvider::addEdgeToBuilder(Step::Edge const& edge,
+                                       arangodb::velocypack::Builder& builder) {
+  insertEdgeIntoResult(edge.getID(), builder);
+};
+
+void ClusterProvider::Step::Edge::addToBuilder(ClusterProvider& provider,
+                                               arangodb::velocypack::Builder& builder) const {
+  provider.insertEdgeIntoResult(getID(), builder);
+}
+
+ClusterProvider::ClusterProvider(arangodb::aql::QueryContext& queryContext,
+                                 ClusterBaseProviderOptions opts,
+                                 arangodb::ResourceMonitor& resourceMonitor)
+    : _trx(std::make_unique<arangodb::transaction::Methods>(queryContext.newTrxContext())),
+      _query(&queryContext),
+      _resourceMonitor(&resourceMonitor),
+      _opts(std::move(opts)),
+      _stats{} {
+  _cursor = buildCursor();
+}
+
+std::unique_ptr<RefactoredClusterEdgeCursor> ClusterProvider::buildCursor() {
+  return std::make_unique<RefactoredClusterEdgeCursor>(trx(), _opts.getExpressionContext(),
+                                                       _opts.getCache(),
+                                                       _opts.isBackward());
+}
+
+auto ClusterProvider::startVertex(VertexType vertex) -> Step {
+  LOG_TOPIC("78156", TRACE, Logger::GRAPHS) << "<ClusterProvider> Start Vertex:" << vertex;
+
+  // Create default initial step
+  // Note: Refactor naming, Strings in our cache here are not allowed to be removed.
+  LOG_DEVEL << "Cluster Provider set startVertex to:" << vertex.toString();
+  return Step(_opts.getCache()->persistString(vertex));
+}
+
+auto ClusterProvider::fetch(std::vector<Step*> const& looseEnds)
```

Comment:
Do we want to move that whole logic somewhere else? (previously used in bloated ClusterMethods)